Shopify/cli
# `tunnel` flag and default localhost

### WHY are these changes introduced?
As part of the new `dev` flow, we want `localhost` to be the default option but with tunnels being easy to opt-in.
This is the first iteration. Tunnel selection and multiple plugins will come later.

Fixes https://github.com/Shopify/shopify-cli-planning/issues/326



### WHAT is this pull request doing?
- Add to `tunnel` flag (false by default)
- Make `tunnel`, `no-tunnel` and `tunnel-url` incompatible
- Make `no-tunnel` public (and false by default)
- Refactor some `dev` code for better readability
- Extract frontend URL generation and test it



### How to test your changes?
- Create an app without extensions
  - `dev` that app, you should see `localhost` links
  - use `dev --tunnel`, you should see ngrok links
  - use both `--tunnel` and `--tunnel-url` you should see an error
- Add an extension to the app:
  - `dev` without `tunnel`, you should see ngrok links
  - `dev` with `no-tunnel` you should see localhost links


### Post-release steps

Release this documentation PR: https://github.com/Shopify/shopify-dev/pull/25786

### Measuring impact

How do we know this change was effective? Please choose one:


### Checklist
